dbms/src/Storages/Page/V3/Universal/RaftDataReader.cpp

```diff
@@ -37,18 +37,31 @@ void RaftDataReader::traverse(const UniversalPageId & start, const UniversalPage
     for (const auto & page_id : page_ids)
     {
         const auto page_id_and_entry = uni_ps.page_directory->getByID(page_id, snapshot);
-        acceptor(page_id, uni_ps.blob_store->read(page_id_and_entry));
+        const auto & checkpoint_info = page_id_and_entry.second.checkpoint_info;
+        if (checkpoint_info.has_value() && checkpoint_info.is_local_data_reclaimed)
+        {
+            acceptor(page_id_and_entry.first, uni_ps.remote_reader->read(page_id_and_entry));
+        }
+        else
+        {
+            acceptor(page_id_and_entry.first, uni_ps.blob_store->read(page_id_and_entry));
+        }
     }
 }
 
-void RaftDataReader::traverseRaftLogForRegion(UInt64 region_id, const std::function<void(const UniversalPageId & page_id, DB::Page page)> & acceptor)
+void RaftDataReader::traverseRemoteRaftLogForRegion(UInt64 region_id, const std::function<void(const UniversalPageId & page_id, const PS::V3::CheckpointLocation & location)> & acceptor)
 {
     auto start = UniversalPageIdFormat::toFullRaftLogPrefix(region_id);
-    auto end = UniversalPageIdFormat::toFullRaftLogPrefix(region_id + 1);
-    traverse(start, end, [&](const UniversalPageId & page_id, const DB::Page & page) {
-        if (page_id.hasPrefix(start))
-            acceptor(page_id, page);
-    });
+    auto end = UniversalPageIdFormat::toFullRaftLogScanEnd(region_id);
```

Comment:
why here must use `toFullRaftLogScanEnd(region_id)` but not `toFullRaftLogPrefix(region_id + 1)`?